### PR TITLE
Update idemp provider to handle block and mnt vols

### PR DESCRIPTION
**Pull request by codenrhoden**
This patch updates the idempotency provider to handle both block and
mount volume types, as they need to be handled a bit differently. This
information is not always given directly in the CSI request, so right
now it is inferred by the target path. If the path is a dir, we assume
mount, if the path is a file, we assume block.

We also simplify the logic of looking for mounts (to see if published or
not) by only matching the source device and the target, rather than also
caring about the intermediate "private" mount in between. We can do this
because the interceptor has been been updated previously to catch any
errors returned from the CSI bridge, and if there are any "dangling"
mounts, we trust our bridge to return an error, which should cause a
retry.

We also updated the CSI bridge to behave directly when the idempotency
interceptor and provider are not in play -- the logic to look for
existing mounts was not in place, but is now.

### CODE REVIEW

**Comment by codenrhoden** [comment]
@akutz this patch also kept the changing of `attTokens` over to a struct, even though it is not necessary. Let me know if you are okay with that.

File: agent/csi/libstorage/csi_service_libstorage.go [hunk 1]
```diff
@@ -292,6 +299,10 @@ func (d *driver) ControllerUnpublishVolume(
 		return nil, err
 	}
 
+	d.pubInfoRWL.Lock()
+	defer d.pubInfoRWL.Unlock()
+	delete(map[string]*pubInfoCache{}, volumeID)
```
Comment (akutz) [inline agent/csi/libstorage/csi_service_libstorage.go]: Hi @codenrhoden,

Shouldn't this line be:

```go
delete(d.pubInfo, volumeID)
```

File: agent/csi/libstorage/csi_service_libstorage.go [hunk 2]
```diff
@@ -448,6 +459,11 @@ func (d *driver) NodePublishVolume(
 	target := req.GetTargetPath()
 	vi := req.GetPublishVolumeInfo()
 
+	st, err := os.Stat(target)
```
Comment (akutz) [inline agent/csi/libstorage/csi_service_libstorage.go]: Hi @codenrhoden,

We already import gotil, so why not use it? Just curious.

File: agent/csi/libstorage/csi_service_libstorage.go [hunk 3]
```diff
@@ -486,10 +502,16 @@ func (d *driver) NodePublishVolume(
 	}
 
 	if mv := req.VolumeCapability.GetMount(); mv != nil {
+		if !st.IsDir() {
```
Comment (akutz) [inline agent/csi/libstorage/csi_service_libstorage.go]: Hi @codenrhoden,

Nevermind, I see why you didn't use gotil. You're using the result of the stat call here. Thanks.

**Comment by akutz** [comment]
Hi @codenrhoden,

Because of the required change to the `delete` call, if you want to go ahead and change back the attach token cache from structs to bools, that's fine. Otherwise I wouldn't bother. But since a change is required anyway...

File: agent/csi/libstorage/csi_service_libstorage.go (hunk 1 above)
Comment (codenrhoden) [reply agent/csi/libstorage/csi_service_libstorage.go]: probably. I think I got that from Atom's autofill.  Sad, but I've never had to delete an item from a map before. :)  I can make that change.

File: agent/csi/libstorage/csi_service_libstorage.go (hunk 2 above)
Comment (codenrhoden) [reply agent/csi/libstorage/csi_service_libstorage.go]: yep, exactly. I started with `gotil` as requested, but then noticed I was going to use `os.Stat` to find out if it is a dir anyways...

**Comment by codenrhoden** [comment]
@akutz, yeah, I could go either way on the token struct. I'm inclined to leave it, as it allows for other things to be added if needed. But it is also unlikely to be needed, as that would imply some addition to libStorage across all drivers...

I'll leave it unless you say no. :)

File: agent/csi/libstorage/csi_service_libstorage.go (hunk 1 above)
Comment (akutz) [reply agent/csi/libstorage/csi_service_libstorage.go]: Let me restate it -- My suggestion wasn't really a suggestion. I was being polite :)

**Comment by akutz** [comment]
Hi @codenrhoden,

I'm with you -- let's leave the cache in case we need it for something else in the future.

File: agent/csi/libstorage/csi_service_libstorage.go (hunk 1 above)
Comment (codenrhoden) [reply agent/csi/libstorage/csi_service_libstorage.go]: thanks for catching this. silly mistake

Comment (akutz) [reply agent/csi/libstorage/csi_service_libstorage.go]: No worries. Good catch on the Atom autofill doing the wrong thing. That's weird. And I guess I see why no linter or compiler caught it. You're technically deleting a key/value pair from a map. It's just a brand new map :)

**Pull request merged**